cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/ProgramRunDispatcherInfo.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.deploy.pipeline;
+
+import io.cdap.cdap.app.program.ProgramDescriptor;
+import io.cdap.cdap.app.runtime.ProgramOptions;
+import org.apache.twill.api.RunId;
+
+/**
+ * Information required by {@link io.cdap.cdap.app.deploy.ProgramRunDispatcher} to execute
+ * program-run logic.
+ */
+public class ProgramRunDispatcherInfo {
+
+  private final ProgramDescriptor programDescriptor;
+  private final RunId runId;
+  private final ProgramOptions programOptions;
+
+  /**
+   isDistributed is required to know which implementation of ProgramRuntimeService is to be used.
```

Comment:
this isn't true is it? From your explanation it seemed like this is only used when copying artifacts?

Reply:
That's correct. There are 2 implementations of ProgramRuntimeService, the only difference from Program Run perspective is for DistributedProgramRuntimeService we perform copyArtifacts under an impersonator; and the getExtraParameters function for InMemoryProgramRuntimeService returns a non-empty map(Hostname mapping).

Reply:
We've captured both the cases with the isDistributed flag in the respective implementations in InMemoryProgramRunDispatcher.

Reply:
Remove this comment.

This class shouldn't know anything about multiple implementations of some other class. AbstractProgramRuntimeService sets this flag, which indicates that the program whether the run is distributed or not. What something else does with the flag is not the business of this class.

I still think the way this is used is fairly hacky but we can revisit afterward.